src/osd/PrimaryLogPG.cc

```diff
@@ -2969,6 +3042,171 @@ void PrimaryLogPG::do_proxy_write(OpRequestRef op, const hobject_t& missing_oid,
   in_progress_proxy_ops[soid].push_back(op);
 }
 
+void PrimaryLogPG::do_chunk_op(OpRequestRef op, const hobject_t& missing_oid, 
+			       ObjectContextRef obc, bool write_ordered)
+{
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  OSDOp *osd_op = NULL;
+  for (unsigned int i = 0; i < m->ops.size(); i++) {
+    osd_op = &m->ops[i];
+    uint64_t cursor = osd_op->op.extent.offset;
+    uint64_t op_length = osd_op->op.extent.offset + osd_op->op.extent.length;
+    uint64_t chunk_length = obc->obs.oi.manifest.chunk_map[0].length;
+    uint64_t padding;
+    assert(chunk_length);
+    while (cursor < op_length) {
+      padding = cursor % chunk_length;
+      padding = chunk_length - padding;
+      uint64_t next_length =  padding;
+      if (cursor + next_length > op_length) {
+	next_length = op_length - cursor;
+      }
+      if (op->may_write() || write_ordered) {
+	do_chunk_write(op, obc, i, cursor, next_length);
+      } else {
+	do_chunk_read(op, obc, i, cursor, next_length);
+      }
+      cursor += chunk_length;
+    }
+  } 
+}
+
+void PrimaryLogPG::do_chunk_write(OpRequestRef op, ObjectContextRef obc, int op_index,
+				  uint64_t req_offset, uint64_t req_length)
+{
+  // NOTE: non-const because ProxyWriteOp takes a mutable ref
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  SnapContext snapc(m->get_snap_seq(), m->get_snaps());
+  object_manifest_t *manifest = &obc->obs.oi.manifest;
+  uint64_t chunk_length = manifest->chunk_map[0].length;
+  uint64_t chunk_index = req_offset - (req_offset % chunk_length);
+
+  if (!manifest->chunk_map.count(chunk_index)) {
+    hobject_t ori_soid = m->get_hobj();
+    string chunk_name = ori_soid.oid.name;
+    chunk_name = chunk_name + "_" + to_string(chunk_index);
+    manifest->chunk_map[chunk_index] = manifest->chunk_map[0];
+    manifest->chunk_map[chunk_index].oid.oid = object_t(chunk_name);
+    manifest->chunk_map[chunk_index].flags = chunk_info_t::CHUNK_CLEAN;
```

Comment:
Either way, for a minimal next step, maybe we simply implement the chunked *read* and ignore the chunked write case?  A simplistic approach would be for a write to a chunked object to trigger a promote and then a local write.